src/pq_flash_index.cpp

```diff
@@ -478,6 +502,41 @@ inline int32_t PQFlashIndex<T, LabelT>::get_filter_number(const LabelT &filter_l
     return idx;
 }
 
+template <typename T, typename LabelT>
+void PQFlashIndex<T, LabelT>::generate_random_labels(std::vector<LabelT> &labels, const uint32_t num_labels,
+                                                     const uint32_t nthreads)
+{
+    std::random_device rd;
+    labels.clear();
+    labels.resize(num_labels);
+
+    uint64_t num_total_labels =
+        _pts_to_label_offsets[num_points - 1] + _pts_to_labels[_pts_to_label_offsets[num_points - 1]];
+    std::mt19937 gen(rd());
+    std::uniform_int_distribution<uint64_t> dis(0, num_total_labels);
+
+    tsl::robin_set<uint64_t> skip_locs;
+    for (uint32_t i = 0; i < num_points; i++)
+    {
+        skip_locs.insert(_pts_to_label_offsets[i]);
+    }
+
+#pragma omp parallel for schedule(dynamic, 1) num_threads(nthreads)
+    for (uint32_t i = 0; i < num_labels; i++)
+    {
+        bool found_flag = false;
+        while (!found_flag)
```

Comment:
considering that the average of label number for each node is small, so this way of sampling may not be efficient. You could count the frequency of each label, then use std::discrete_distribution to generate the random value based on the label's frequency.

Reply:
or you also could consider to save the label when we generate the sample data after building index

Reply:
and do you consider to generate the cache node list and save it to some file after building index. Actually, we do not like the extra time cost to search and generate the cache node list during loading index.

Reply:
trouble with generating cache list at build time is that we need to run sample searches to count the most frequent hits. For large index, the search will have to be a disk-index search, so we will add a dependency that build time also needs a good SSD. That is why we went with this design.

Reply:
Hmm, counting label frequency and then sampling, is similar to sampling from the "expanded filter list", which is just a concatenated vector of all the labels right, except we need to discount the elements storing the offsets. If you can explain to me, or find examples where it is much slower, we can consider re-implementing.

Reply:
we use the same hardwares as our prod environments. so we could search the index during building index. Not sure if you have concerns to add a parameter in the build_disk_index.cpp to get the cache list

Reply:
> Hmm, counting label frequency and then sampling, is similar to sampling from the "expanded filter list", which is just a concatenated vector of all the labels right, except we need to discount the elements storing the offsets. If you can explain to me, or find examples where it is much slower, we can consider re-implementing.

The final result is the same. I do not have the performance result. just feel that the number of unique labels is much smaller than the total number of labels. and the implementation is much simpler. I am fine if you want to keep current implementation.